Brings EvaporateJS to Version 1.0.0

bikeath1337 opened the pull request:
This pull request uses the existing EvaporateJS coding standards and  patterns to facilitate review. As such there is still a lot of code that could be improved by applying DRY. This PR leaves that improvement for the future.

Note that the CORS configuration and S3 Bucket policies need to be updated to take advantage of the new features. Fortunately, these changes are backward compatible with previous versions, so they should be addressed before deploying this version.

Assuming this passes muster, I believe we can call this a stable release @ 1.0.0.

# Breaking Change
In order to allow resumption of uploads without re-uploading what has already been sent to AWS, the `complete` callback needed to be updated to receive the AWS S3 object key of the remote resource that matches what the client is trying to upload. In this case, the client's proposed S3 object key is not used since the object is already present onS3.

The following issues are addressed in this PR:

1. Issue #96 Support for Md5 Digests
2. Issue #98 Support for aborting multipart uploads
3. Issue #100 Correct memory leak due to infinity retry feature
4. Issue #104 Ability to resume uploads, even after client failure or browser refresh
5. Issue #106 Corrects documentation around the .add() method.
6. Issue #107 Adds ability to retry failures when initiating and completing a multipart upload (required by AWS)

tomsaffell commented:
@bikeath1337 thanks for doing this! very much simplifies reviewing. I'll review and test this week, then hopefully we can get it merged into master!

bikeath1337 commented:
No problem. Very important to make sure the you have your access policy configured for your S3 user and also that you've updated your S3 CORS configuration to allow the new methods. If you don't do the CORS config and CORS is triggered, you'll see OPTIONS requests fail with a 400 type error.

If you fail to configure the bucket access policy, you'll see 400 type errors on the requests to S3.

cnorthwood commented on evaporate.js:
```diff
@@ -14,25 +14,27 @@ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 
-     if(!this.supported){
+     if (!this.supported) {
+        l.e('The broswer does not support the necessary features of File and Blob [webkitSlice || mozSlice || slice || readAsBinaryString]')
```
l isn't actually defined until later down in this function, so calling it at this point you get a lovely `Uncaught TypeError: Cannot read property 'e' of undefined` in the console (plus there's a typo)

bikeath1337 replied:
Thanks. I added that part at the very end, so got sloppy. Will fix now.

cnorthwood commented on evaporate.js:
```diff
@@ -14,25 +14,27 @@ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+           Blob.prototype.webkitSlice ||
+           Blob.prototype.mozSlice ||
+           Blob.prototype.slice ||
+           (config.computeContentMd5 && Blob.prototype.readAsBinaryString) === 'undefined') ||
```
I think this test is wrong. `readAsBinaryString` is method on FileReader, not Blob (currently the latest Chrome fails this test), and it's allegedly been replaced by `readAsArrayBuffer` in the spec

cnorthwood replied:
I think this test wants to be:

```
     this.supported = !(
        typeof File === 'undefined' ||
        typeof Blob === 'undefined' ||
        typeof (
           Blob.prototype.webkitSlice ||
           Blob.prototype.mozSlice ||
           Blob.prototype.slice) === 'undefined' ||
        (!!config.computeContentMd5 && typeof FileReader.prototype.readAsBinaryString === 'undefined') ||
           !!config.testUnsupported)
```

bikeath1337 replied:
Yes, correcting that now as well. lunch break.

cnorthwood replied:
thanks very much :)

bikeath1337 replied:
Regarding `readAsBinaryString`. I did not encounter that warning and was following what has worked and is in the code. The latest 'supported' check corrects the reference from Blob to FileReader.

Let me replace this with readAsArrayBuffer and see where it goes.

bikeath1337 commented:
I wanted to avoid refactoring this a while ago but alas. Thanks for the comments. As I wrote earlier, I added this after the rest of the code was completed and thus it has had little scrutiny.

Just re-arranged the validation checks with my latest commit.

bikeath1337 commented:
The last commit was an attempt to get logging for the specific startup scenario which should not obey logging restrictions. There is a lot of refactoring we can do here but I think it's best to address this in a later phase.

bikeath1337 commented:
On my test branch, I noticed that after Cancel, the checksum calculations continue but they actually should end. I will look into that later.

bikeath1337 commented:
@tomsaffell I think we should add a line or two to the README to let visitors know there's activity here. What do you think about that? We could just push the text change to master and then people would see activity on the timestamps and see it in the readme.

tomsaffell commented:
@bikeath1337 updating readme sounds good. that said, im doing some testing now, so if all goes well we may be able to merge this branch into master today.

tomsaffell commented:
@bikeath1337 just looking at the new S3 Bucket Policy. Maybe I'm misreading it, but it's not clear to me how one would configure the permissions to that a client can list the bucket for just its files w/o also allowing it to list all files for all clients. Maybe I'm missing something. If you're available now, want to Skype? I'm tom.saffell

cnorthwood commented:
I think that there needs to be additional logic on the side of the server to be more careful about which requests it should sign, but that shouldn't block merging the JS side, just needs more documentation. It's something I plan to look at next week.

tomsaffell commented:
@bikeath1337 just tested the 'resume after page reload' (RAPR) feature - it's fantastic - nice work. We need a good name for it, to distinguish from the normal resuming. Any suggestions better than RAPR? Regarding the bucket policy and security...

@cnorthwood yes, additional logic on the server side would be a good way to solve this, I think that is a matter of documentation! That said, I think it's fairly easy to lock it down without that. AFAICT, the RAPR feature only needs the following permission:

```
{
	"Version": "2012-10-17",
	"Id": "Policy145337ddwd",
	"Statement": [
		{
			"Sid": "",
			"Effect": "Allow",
			"Principal": {
				"AWS": "xxxxx"
			},
			"Action": [
				"s3:ListMultipartUploadParts"
			],
			"Resource": "arn:aws:s3:::mybucket/*"
		}
	]
}
```
The request is

`https://s3.amazonaws.com/mybucket/key_name?uploadId=xyz&part-number-marker=0`

The uploadId, if I understand AWS right, is a random ID that is only known to the client who first tried the upload - does that sounds right, or is uploadId guessable? Also, the key_name could easily be an obfuscation. And of course the whole request has to be signed server side. So I believe that this is actually safe thing to do, and does not allow one user to access data relating to another user. @bikeath1337 Do you agree?

@bikeath1337 please could you add to the documentation details about exactly which permissions are required for each of the features

For example, are these two permissions only needed for periodic cleanup of incomplete uploads?
```

"Action": [
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads"
            ],
            "Resource": "arn:aws:s3:::mybucket"
```

If so, and if I'm understanding correctly that this feature is not implemented in this version of EvaporateJS (i.e. it's just being recommended as good housekeeping) then I suggest we dont put that in the 'template' policy - but maybe I misunderstood what it's there for?

tomsaffell commented:
@bikeath1337 for the sake of documentation: I tested the Md5 feature on Chrome, and it works very nicely. Does not seems to consume lots of memory, nor CPU. 1500ms pause is definitely needed though to prevent CPU pegging :)

tomsaffell commented:
@bikeath1337 I tested the bit that calls http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadAbort.html

It seems to work, and the `checkForParts` is a nice touch. 

We might want to give the `cancel()` method an optional paramter `dontAbortParts`, so that it can be implemented in a way that allows a user to stop BW usage and still be able to restart later. But maybe that's just `pause`, depending on how the user is thinking about it. Anyway, it's not for now.

tomsaffell commented:
@bikeath1337 I have some suggestions for the README. I think what I'll do is once we have this merged into master I'll make those in a separate commit direct to Master. It's mostly just about clarifying a few things (e.g. difference between resuming after page reload vs 'normal' resuming)

tomsaffell commented:
@bikeath1337 I'm now done with my 'user' testing, and it all seems to work as expected. I've also reviewed the code, and it *looks* right (and yes, a refactor would be a nice next step). I havent run any rigorous tests, because I dont have any ;). So with all that said, except for possibly changing the recommended bucket policy, I think it's ready for master. Thanks again - this is a big leap forward.

SamPotts commented:
This looks good. Keen to see this get released.

SamPotts commented:
@cnorthwood: Security is probably something to be discussed elsewhere but there is definitely an issue with just signing any request. Without giving too much away, we built a token system for such a thing; a user is issued a token (based on what bucket/folder they are uploading too) before any upload takes place and that token must match when signing requests. We also move files from temporary buckets after processing (using AWS lambda) which helps mitigate the risk. I noticed a lot of sites implementing CORs that do not have such security in place and leaving themselves open to having assets overwritten pretty easily by simply using developer tools.

bikeath1337 commented:
@cnorthwood regarding the server side signing endpoint...that's the responsibility of the developer who's using EvaporateJS. I think that it would be better served to point to AWS documentation about how to sign URLs...even though the documentation is fragmented or all over the place.

I do agree with @tomsaffell that we need to look at the S3 policy and make sure we understand the ramifications and therefore advise end users of risks and assumptions.

bikeath1337 commented:
@tomsaffell I don't mind RAPR except that it's awfully close to "reaper" and "raper" so we might want a less charged term. LOL.

RACE: Resume After Client Error?

bikeath1337 commented:
@tomsaffell I added a commit to this PR to try to describe the issues around the RACE feature and S3 permissions.

SamPotts commented:
@bikeath1337 I don't think an S3 bucket policy will prevent the issue I mentioned above. I'm not sure how uploading through signed urls would work (if it does).

bikeath1337 commented:
@SamPotts there are two threads here: 1) S3 policy required for RARE and 2) note about security in the signing url implementation, AFAIK. The two are not directly related.

SamPotts commented:
Ah okay, no worries. RARE/RACE/RAPR being the resume after reload?

bikeath1337 commented:
Yes.

tomsaffell commented:
@bikeath1337 thanks. Agreed that the issue of general signing security/warnings is separate from the issue of the policy for the RACR (resume after client restart, pronounced 'racer'). 

Thanks for updating the docs - looks like the readme is still suggesting a bucket policy that allows the `ListBucket` - what function is that necessary for? From reading the code and testing, it seems none, but maybe i missing something. If we dont need it, i suggest we get rid of it - it's a pretty big thing to grant. Sorry if you already addressed that and I missed it.

bikeath1337 commented:
Yeah, that might have been a relic from when I was trying to get it to work. Did you try it without the permission? I'll do so in the morning.

bikeath1337 commented:
I confirmed that the listBucket permission is not required. It was a relic left over from my development.

bikeath1337 commented:
@tomsaffell , I added a new configuration parameter which enables a developer to disable the extra check against S3. This puts control in their own hands. The default is to disable the feature, so the S3 bucket permissions do not require the `s3:GetObject` permission.

I also fixed some relics in our options formatting.

bikeath1337 commented:
Finally, I updated the documentation to remove references to requiring the HEAD method in the CORS settings if the S3 optimization is disabled since the uploader will issue no HEAD request.

Heh, heh...I think we're ready?

tomsaffell commented:
Thanks @bikeath1337. Sorry, I should have asked this before, but is the `ListBucketMultipartUploads` policy actually needed? I dont think it is (I did my testing w/o it, and it seems the code never makes such a request)

> a relic from when I was trying to get it to work

hehe, that was my guess ;) I always end up with a couple of those..

bikeath1337 commented:
Agreed, @SamPotts ! That's why we collaborate like this. 

Yes it is not needed. I copied my policy and forgot that I added the actions to support the orphaned multi-part cleanup, which, of course, requires that action. I will remove it now.

Thanks again.

bikeath1337 commented:
Removed references to the unneeded action.

tomsaffell commented:
Great. Thanks @bikeath1337 I think we're ready! You want to do the honors? You have the github permissions, right?

SamPotts commented:
Awesome. I can't wait to give this a go! :+1: Nice work.

bikeath1337 commented:
Alrighty then, I'm going to tag this as `1.0.0` and fast-forward master to HEAD of `v1`.

Thank you all for your productive collaboration.

tomsaffell commented:
Great! Thanks again @bikeath1337 this is 99% your work!

SamPotts commented:
Awesome work :+1:

cnorthwood commented:
Cheeky request, any chance of getting v1 pushed to NPM? :)

bikeath1337 commented:
Yes, will do. Thank you.
